### PR TITLE
Oracle support (yes, really)

### LINKED ISSUE #345: Explore Knex Oracle support
Because I am absolutely insane

### PR DESCRIPTION
**Pull request by davidmckenzie**
# Description

Adds support for OracleDB, also includes optimisations to the search code to make it more knexy

Fixes #345 

Note: This probably needs a lot of documentation. I forget most of the stuff I had to do just to get the OracleDB to connect, it was annoying. Luckily I have pages of scribbled notes that make no sense.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Tested on SQLite and OracleDB, have not tested on MySQL

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated changelog.md with changes made

### CODE REVIEW

**Comment by DanrwAU** [comment]
I don't have an oracle test environment, but i'll run through tomorrow and run the SQLite and MySQL tests i ran last time and ensure nothing new is broken.

**Comment by davidmckenzie** [comment]
Thanks dude :) Yeah I'm not expecting _anyone_ to use the oracle side, so review is more making sure it doesn't break your stuff. Have tested pretty extensively on sqlite and looks good there (my prod is running on this branch)

**Comment by DanrwAU** [comment]
First issue, 

It's not passing credentials correctly to MySQL to open the database. 

`
[pmon]  2020-01-26 01:13:12  error : Error upgrading database:Error: ER_ACCESS_DENIED_ERROR: Access denied for user ''@'localhost' (using password: NO)
`

image

File: server/knexfile.js [hunk 1]
```diff
@@ -40,6 +36,21 @@ var dbconfig = {
       },
     }
 }
+
+if (dbtype == 'sqlite3') {
+  dbconfig.connection.filename = nconf.get('database:file');
+} else if (dbtype == 'sqlite3') {
```
Comment (DanrwAU) [inline server/knexfile.js]: This should be 

`
else if (dbtype == 'mysql')
`

**Comment by DanrwAU** [comment]
Searching is also broken in MySQL 

`
6|pagermon-test  | [pmon]  2020-01-26 01:19:53  error : select `messages`.*, `capcodes`.`alias`, `capcodes`.`agency`, `capcodes`.`icon`, `capcodes`.`color`, `capcodes`.`ignore` from `messages` left join `capcodes` on `capcodes`.`id` = `messages`.`alias_id` MATCH(messages.message, messages.address, messages.source) AGAINST (''hawthorn'' IN BOOLEAN MODE) order by `messages`.`timestamp` desc - ER_PARSE_ERROR: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'MATCH(messages.message, messages.address, messages.source) AGAINST (''hawthorn''' at line 1
`

**Comment by DanrwAU** [comment]
I may have found a way to fix full text searching in MySQL while trying to fix searching in this... Happy for me to add this change to line 273 @davidmckenzie?
` query = '"' + query + '"' `

Forces MySQL to search for the string rather than the individual words, which has been a problem since i implemented it. Knex seems to do funky things with quotes when they aren't part of the var. 
I've checked and it doesn't seem to be SQL Injectionable so double win!

**Comment by davidmckenzie** [comment]
👌 thanks dude, yah that looks good!

**Comment by DanrwAU** [comment]
@davidmckenzie I don't remember are we happy with this one now? 🤔

**Comment by davidmckenzie** [comment]
@DanrwAU yep I'm happy as long as you're happy with the MySQL side!

**Pull request merged**